drivers/ethernet/eth_mcux.c

```diff
@@ -563,6 +563,16 @@ static void generate_mac(u8_t *mac_addr)
 	/* Locally administered, unicast */
 	mac_addr[5] = ((entropy >> 0) & 0xfc) | 0x02;
 }
+#elif defined(CONFIG_ETH_MCUX_0_UNIQUE_MAC)
+static void generate_mac(u8_t *mac_addr)
+{
+	u32_t id = SIM->UIDL;
+
+	mac_addr[3] = id >> 8;
```

Comment:
This uses only 22 bits of the 80 bit unique number.  My understanding of the comments from Maureen in the original ticket is that the number is not unique unless we draw on all 80 bits. There was previously some discussion  about hashing the 80bit unique ID and drawing the 22 bits we need from the hash... is there a reason you've chosen not to do this?

Reply:
> This uses only 22 bits of the 80 bit unique number.

Yes, because we need 22 unique bits. So, no matter how many unique bits we'd have, we'll use only 22 of them.

> My understanding of the comments from Maureen in the original ticket is that the number is not unique unless we draw on all 80 bits.

That's not impression I got. She said that there're 80 bits in the unique identifier (and looking at MCUX HAL code, for some other MCUX MCUs, there can be different number of bits), that's all.

> There was previously some discussion about hashing the 80bit unique ID and drawing the 22 bits we need from the hash... is there a reason you've chosen not to do this?

I actually implemented the hashing - the trivial one, extract_22_bits_from(long_value). The discussion you mentioned evolved around "because we can" argument (I'd even say it was a typical programmers' smalltalk on a trivial, but not surmountable matter - hash functions). Nobody gave arguments why anything more than the trivial hash function is required.

And this patch is to resolve the very specific issue of MAC address assignment, not to flex muscles in implementing a-bit-more-trivial-than-completely-trivial hash functions. This patch solves the intended problem, and expectedly does that with minimal changes/code/overhead. If (when) someone can prove there's a problem with how it does it, it can be easily improved. (E.g. the next step will be xoring different words of uid together.)

Reply:
> We have no idea what the distribution of values is in the 22 bits this patch has chosen to use.

Exactly, we have no idea about the distribution, but you somehow assume that it is "bad".

> Since a hash of 80 bits is trivial

Hash of 80 bits is trivial, and implemented here. Any a bit less trivial hash is no better than that, except by speculation. The problem with this path is that there's no criteria where to stop. For example, next guy joining the discussion will very eloquently tell that the only good hash function is SHA512, anything "below" it is bad, bad, because ... . So, the place to stop should be at the very beginning, with the simplest solution which works. If it's proven to be not good enough, next candidate can be used (which also can be disproved, all hash functions have collisions you know). In the meantime, effort of inventing magic hash functions and/or making hardware drivers depend on libraries unrelated to their functionality should be stopped, and the effort invested in the real problem spots (which are many).

Reply:
Indeed, 'bad' because if multiple device on my network have the same 22 bits in the 80 bit id they would end up with the same MAC address.  Since we have no idea what the allocation policy is in the manufacturer we have no idea how many devices share the same 22 bits being used here.

"Hash of 80 bits is trivial, and implemented here."

I'm not sure I follow your point here.  Clearly what is implemented here makes no attempt whatsoever to use more than the least significant 22 bits of uidl...

Reply:
> Indeed, 'bad' because if

Again, that "if". I have no idea what makes you think that you'd get duplicate MAC address. Did you actually test it? If not, why so much worrying? Why this desire to over-insure? (This all is devel/debugging feature. For production, a vendor will need to assign a MAC address directly owned by them anyway.)

Anyway, @mjs-arm, you asked a question why it was done that way. I did explain it in details and from different sides. At the same time, you gave it -1, while didn't provide specific suggestions (with arguments) on improving it. Anyway, I made a change to use all 128 (sic!) bits of MCUX's unique MCU identifier. I hope you'll be satisfied now.